### PR TITLE
Xkqian/pr/add rsa pss rsae

### PR DESCRIPTION
**Pull request by xkqian**
Add RSA signature schemes PSS in certificate verify.

### CODE REVIEW

File: tests/ssl-opt.sh [hunk 1]
```diff
@@ -8837,6 +8837,36 @@ run_test    "TLS1.3: Test client hello msg work - openssl" \
             -c "<= parse certificate verify"          \
             -c "mbedtls_ssl_tls13_process_certificate_verify() returned 0"
 
+requires_openssl_tls1_3
```
Comment (ronald-cron-arm) [inline tests/ssl-opt.sh:8840]: As mentioned offline, for testing server authentication based on RSA_PSS_RSAE_SHA256, look in ssl-opt.sh in the prototype branch (https://github.com/hannestschofenig/mbedtls) the tests with a header containing "Server authentication with RSA_PSS_RSAE_SHA256".

Comment (xkqian) [reply tests/ssl-opt.sh:8840]: Thanks a lot. Will check that.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: library/ssl_tls.c [hunk 2]
```diff
@@ -6370,6 +6370,12 @@ static uint16_t ssl_preset_default_sig_algs[] = {
     MBEDTLS_TLS13_SIG_ECDSA_SECP521R1_SHA512,
 #endif /* MBEDTLS_SHA512_C && MBEDTLS_ECP_DP_SECP521R1_ENABLED */
 #endif /* MBEDTLS_ECDSA_C */
+
+    /* RSA algorithms */
+#if defined(MBEDTLS_X509_RSASSA_PSS_SUPPORT)
+    MBEDTLS_TLS13_SIG_RSA_PSS_RSAE_SHA256,
```
Comment (yuhaoth) [inline library/ssl_tls.c:6376]: Name should follow up name schema. to align with #5166 
```suggestion
    MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA256,
```

File: library/ssl_tls.c [hunk 3]
```diff
@@ -6383,6 +6389,12 @@ static uint16_t ssl_preset_suiteb_sig_algs[] = {
     MBEDTLS_TLS13_SIG_ECDSA_SECP384R1_SHA384,
 #endif /* MBEDTLS_SHA512_C && MBEDTLS_ECP_DP_SECP384R1_ENABLED */
 #endif /* MBEDTLS_ECDSA_C */
+
+    /* RSA algorithms */
+#if defined(MBEDTLS_X509_RSASSA_PSS_SUPPORT)
+    MBEDTLS_TLS13_SIG_RSA_PSS_RSAE_SHA256,
```
Comment (yuhaoth) [inline library/ssl_tls.c:6395]: ditto

File: tests/ssl-opt.sh [hunk 4]
```diff
@@ -8835,6 +8841,22 @@ run_test    "TLS1.3: minimal feature sets - openssl" \
             -c "<= parse finished message" \
             -c "HTTP/1.0 200 ok"
 
+requires_openssl_tls1_3
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
+requires_config_enabled MBEDTLS_DEBUG_C
+requires_config_enabled MBEDTLS_SSL_CLI_C
+requires_config_enabled MBEDTLS_X509_RSASSA_PSS_SUPPORT
+requires_config_disabled MBEDTLS_USE_PSA_CRYPTO
+run_test    "TLS 1.3 m->O AES_128_GCM_SHA256      , RSA_PSS_RSAE_SHA256" \
```
Comment (yuhaoth) [inline tests/ssl-opt.sh:8850]: Please keep this conversion. It should be moved to `compat-tls13-testcases.sh` in #5212 
No action needed in this PR

**Comment by yuhaoth** [comment]
Please add new component  in `all.sh`

File: library/ssl_tls13_generic.c [hunk 5]
```diff
@@ -368,6 +373,13 @@ static int ssl_tls13_parse_certificate_verify( mbedtls_ssl_context *ssl,
             md_alg = MBEDTLS_MD_SHA512;
             sig_alg = MBEDTLS_PK_ECDSA;
             break;
+#if defined(MBEDTLS_X509_RSASSA_PSS_SUPPORT)
+        case MBEDTLS_TLS13_SIG_RSA_PSS_RSAE_SHA256:
```
Comment (yuhaoth) [inline library/ssl_tls13_generic.c:377]: How about add `rsa_pss_rsae_sha384` and `rsa_pss_rsae_sha512`?
I do not think those need a new PR

File: tests/ssl-opt.sh [hunk 6]
```diff
@@ -8867,6 +8889,29 @@ run_test    "TLS1.3: minimal feature sets - gnutls" \
...
+requires_gnutls_next_disable_tls13_compat
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
+requires_config_enabled MBEDTLS_DEBUG_C
+requires_config_enabled MBEDTLS_SSL_CLI_C
+requires_config_enabled MBEDTLS_X509_RSASSA_PSS_SUPPORT
+requires_config_disabled MBEDTLS_USE_PSA_CRYPTO
+requires_gnutls_next
+run_test    "TLS 1.3 m->G AES_128_GCM_SHA256      , RSA_PSS_RSAE_SHA256" \
+            "$G_NEXT_SRV_RSA --disable-client-cert --priority=NORMAL:+CIPHER-ALL:+SHA256:+GROUP-SECP256R1:+ECDHE-ECDSA:+AEAD:+SIGN-RSA-PSS-RSAE-SHA256:-VERS-ALL:+VERS-TLS1.3:%NO_TICKETS:%DISABLE_TLS13_COMPAT_MODE" \
+            "$P_CLI debug_level=4 force_version=tls1_3 server_name=localhost force_ciphersuite=TLS1-3-AES-128-GCM-SHA256" \
```
Comment (yuhaoth) [inline tests/ssl-opt.sh:8902]: `sig_algs` parameter of `ssl_client2` should be changed also.

**Review by yuhaoth — CHANGES_REQUESTED** [review]

File: library/ssl_tls.c (hunk 2 above)
Comment (xkqian) [reply library/ssl_tls.c:6376]: Cause all of the macros in ssl_preset_default_sig_algs are TLS1_3, and we leave all of the related tasks into #4998, let's keep it here.  And if need, I can create another PR to change this type.

File: library/ssl_tls.c (hunk 3 above)
Comment (xkqian) [reply library/ssl_tls.c:6395]: Cause all of the macros in ssl_preset_default_sig_algs are TLS1_3, and we leave all of the related tasks into #4998, let's keep it here. And if need, I can create another PR to change this type.

File: tests/ssl-opt.sh (hunk 4 above)
Comment (xkqian) [reply tests/ssl-opt.sh:8850]: Ok

File: library/ssl_tls13_generic.c (hunk 5 above)
Comment (xkqian) [reply library/ssl_tls13_generic.c:377]: We only focus on the rsa_pss_rsae_sha256 in this task, please ignore 384 and 512 currently.

File: tests/scripts/all.sh [hunk 7]
```diff
@@ -2610,6 +2610,22 @@ component_test_tls13_experimental_with_everest () {
     if_build_succeeded tests/ssl-opt.sh
 }
 
+component_test_tls13_experimental_with_rsa () {
+    msg "default config with MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL enabled, with rsa"
+    scripts/config.py   set MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
+    scripts/config.py unset MBEDTLS_SSL_USE_MPS
```
Comment (yuhaoth) [inline tests/scripts/all.sh:2616]:
```suggestion
```

File: programs/ssl/ssl_client2.c [hunk 8]
```diff
@@ -1534,13 +1534,18 @@ int main( int argc, char *argv[] )
             {
                 sig_alg_list[i++] = MBEDTLS_TLS13_SIG_ECDSA_SECP521R1_SHA512;
             }
+            else if( strcmp( q, "rsa_pkcs1_sha256" ) == 0 )
```
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:1537]: Add Macro wrapper

File: programs/ssl/ssl_client2.c [hunk 9]
```diff
@@ -1534,13 +1534,18 @@ int main( int argc, char *argv[] )
...
+                sig_alg_list[i++] = MBEDTLS_TLS13_SIG_RSA_PKCS1_SHA1;
+            }
             else
             {
                 mbedtls_printf( "unknown signature algorithm %s\n", q );
                 mbedtls_printf( "supported signature algorithms: " );
                 mbedtls_printf( "ecdsa_secp256r1_sha256 " );
                 mbedtls_printf( "ecdsa_secp384r1_sha384 " );
                 mbedtls_printf( "ecdsa_secp521r1_sha512 " );
+                mbedtls_printf( "rsa_pkcs1_sha256 " );
```
Comment (yuhaoth) [inline programs/ssl/ssl_client2.c:1548]: ditto

**Review by yuhaoth — CHANGES_REQUESTED** [review]

File: programs/ssl/ssl_client2.c [hunk 10]
```diff
@@ -1534,13 +1534,22 @@ int main( int argc, char *argv[] )
             {
                 sig_alg_list[i++] = MBEDTLS_TLS13_SIG_ECDSA_SECP521R1_SHA512;
             }
+#if defined(MBEDTLS_X509_RSASSA_PSS_SUPPORT)
```
Comment (ronald-cron-arm) [inline programs/ssl/ssl_client2.c:1537]: I do not see here guards for the other signature algorithms thus I'd suggest the same for RSASSA_PSS.

File: tests/scripts/all.sh [hunk 11]
```diff
@@ -2610,6 +2610,20 @@ component_test_tls13_experimental_with_everest () {
     if_build_succeeded tests/ssl-opt.sh
 }
 
+component_test_tls13_experimental_with_rsa () {
```
Comment (ronald-cron-arm) [inline tests/scripts/all.sh:2613]: MBEDTLS_RSA_C and MBEDTLS_X509_RSASSA_PSS_SUPPORT are enabled in the default configuration thus I do not think we need this additional component. `component_test_tls13_experimental` does the job.

File: tests/ssl-opt.sh [hunk 12]
```diff
@@ -8835,6 +8841,22 @@ run_test    "TLS1.3: minimal feature sets - openssl" \
...
 
+requires_openssl_tls1_3
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
+requires_config_enabled MBEDTLS_DEBUG_C
+requires_config_enabled MBEDTLS_SSL_CLI_C
+requires_config_enabled MBEDTLS_X509_RSASSA_PSS_SUPPORT
+requires_config_disabled MBEDTLS_USE_PSA_CRYPTO
+run_test    "TLS 1.3 m->O AES_128_GCM_SHA256      , RSA_PSS_RSAE_SHA256" \
+            "$O_NEXT_SRV_RSA -ciphersuites TLS_AES_128_GCM_SHA256 -tls1_3 -msg -no_middlebox -num_tickets 0" \
+            "$P_CLI debug_level=4 force_version=tls1_3 server_name=localhost force_ciphersuite=TLS1-3-AES-128-GCM-SHA256" \
```
Comment (ronald-cron-arm) [inline tests/ssl-opt.sh:8852]:
```suggestion
            "$P_CLI debug_level=4 force_version=tls1_3 server_name=localhost force_ciphersuite=TLS1-3-AES-128-GCM-SHA256 allow_sha1=0" \
```

File: tests/ssl-opt.sh [hunk 13]
```diff
@@ -8867,6 +8889,29 @@ run_test    "TLS1.3: minimal feature sets - gnutls" \
...
+requires_config_enabled MBEDTLS_DEBUG_C
+requires_config_enabled MBEDTLS_SSL_CLI_C
+requires_config_enabled MBEDTLS_X509_RSASSA_PSS_SUPPORT
+requires_config_disabled MBEDTLS_USE_PSA_CRYPTO
+requires_gnutls_next
+run_test    "TLS 1.3 m->G AES_128_GCM_SHA256      , RSA_PSS_RSAE_SHA256" \
+            "$G_NEXT_SRV_RSA --disable-client-cert --priority=NORMAL:+CIPHER-ALL:+SHA256:+GROUP-SECP256R1:+ECDHE-ECDSA:+AEAD:+SIGN-RSA-PSS-RSAE-SHA256:-VERS-ALL:+VERS-TLS1.3:%NO_TICKETS:%DISABLE_TLS13_COMPAT_MODE" \
+            "$P_CLI debug_level=4 force_version=tls1_3 server_name=localhost force_ciphersuite=TLS1-3-AES-128-GCM-SHA256" \
+            0 \
+            -c "server hello, chosen ciphersuite: ( 1301 ) - TLS1-3-AES-128-GCM-SHA256" \
```
Comment (ronald-cron-arm) [inline tests/ssl-opt.sh:8904]:
```suggestion
```
present below.

File: tests/ssl-opt.sh [hunk 14]
```diff
@@ -8867,6 +8889,29 @@ run_test    "TLS1.3: minimal feature sets - gnutls" \
...
+            -s "Ephemeral EC Diffie-Hellman parameters" \
```
Comment (ronald-cron-arm) [inline tests/ssl-opt.sh:8905]: Please remove the checks of the GnuTLS server logs as when we will change GnuTLS version in the future they may not work anymore and their check does not bring much.

File: tests/ssl-opt.sh (hunk 6 above)
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:8902]:
```suggestion
            "$P_CLI debug_level=4 force_version=tls1_3 server_name=localhost force_ciphersuite=TLS1-3-AES-128-GCM-SHA256 allow_sha1=0" \
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: tests/ssl-opt.sh [hunk 15]
```diff
@@ -8899,16 +8899,9 @@ requires_config_disabled MBEDTLS_USE_PSA_CRYPTO
...
+            "$P_CLI debug_level=4 force_version=tls1_3 server_name=localhost force_ciphersuite=TLS1-3-AES-128-GCM-SHA256 allow_sha1=0" \
             0 \
-            -c "server hello, chosen ciphersuite: ( 1301 ) - TLS1-3-AES-128-GCM-SHA256" \
-            -s "Ephemeral EC Diffie-Hellman parameters" \
-            -s "Version: TLS1.3" \
-            -s "Cipher: AES-128-GCM" \
-            -S "Client Signature:" \
-            -s "Server Signature: RSA-PSS-RSAE-SHA256" \
             -c "ECDH curve: x25519"         \
-            -c "server hello, chosen ciphersuite: ( 1301 ) - TLS1-3-AES-128-GCM-SHA256" \
```
Comment (ronald-cron-arm) [inline tests/ssl-opt.sh]: Please keep this one.

File: programs/ssl/ssl_client2.c (hunk 8 above)
Comment (xkqian) [reply programs/ssl/ssl_client2.c]: I Just change back to align with the exist coding style as comments https://github.com/ARMmbed/mbedtls/pull/5127#discussion_r755114957.  Please help check whether we can close this conversation.

File: programs/ssl/ssl_client2.c (hunk 9 above)
Comment (xkqian) [reply programs/ssl/ssl_client2.c]: I Just change back to align with the exist coding style as comments https://github.com/ARMmbed/mbedtls/pull/5127#discussion_r755114957.  Please help check whether we can close this conversation.

File: tests/scripts/all.sh (hunk 7 above)
Comment (xkqian) [reply tests/scripts/all.sh]: I just remove this component as it has been covered, please see comments: https://github.com/ARMmbed/mbedtls/pull/5127#discussion_r755118046.

**Review by ronald-cron-arm — APPROVED** [review]
LGTM

**Review by yuhaoth — APPROVED** [review]

**Comment by yuhaoth** [comment]
LGTM

**Pull request merged**